src/gridfs/upload.rs

```diff
@@ -173,3 +179,287 @@ impl GridFsBucket {
+                stream.state.set_writing(new_future)
+            }
+            State::Writing(future) => future,
+            State::Closing(_) | State::Closed => return Poll::Ready(Err(get_closed_error())),
+        };
+
+        let result = match future.poll_unpin(cx) {
+            Poll::Ready(result) => result,
+            Poll::Pending => return Poll::Pending,
+        };
+
+        match result {
+            Ok((chunks_written, buffer)) => {
+                stream.current_n += chunks_written;
+                stream.state = State::Idle(Some(buffer));
+                Poll::Ready(Ok(buf.len()))
+            }
+            Err(error) => {
+                stream.state = State::Closed;
+                Poll::Ready(Err(error.into_futures_io_error()))
+            }
+        }
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<tokio::io::Result<()>> {
+        // The buffer only contains leftover bytes that couldn't fill an entire chunk, so there's
+        // nothing to flush.
+        match self.state {
+            State::Closed => Poll::Ready(Err(get_closed_error())),
+            _ => Poll::Ready(Ok(())),
+        }
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<tokio::io::Result<()>> {
+        let stream = self.get_mut();
+
+        let future = match &mut stream.state {
+            State::Idle(buffer) => {
+                let buffer = buffer.take().unwrap();
+
+                let file = FilesCollectionDocument {
+                    id: stream.id.clone(),
+                    length: stream.current_n as u64 * stream.chunk_size as u64
+                        + buffer.len() as u64,
+                    chunk_size: stream.chunk_size,
+                    upload_date: DateTime::now(),
+                    filename: stream.filename.take(),
+                    metadata: stream.metadata.take().unwrap(),
+                };
+
+                let new_future = close(stream.bucket.clone(), buffer, file).boxed();
+                stream.state.set_closing(new_future)
+            }
+            // This case is effectively unreachable, as the AsyncWriteExt methods take &mut self and
+            // poll the futures to completion. If a user were to call these polling methods directly
+            // and intersperse futures, we should just pend here until the writing future resolves.
+            State::Writing(_) => return Poll::Pending,
+            State::Closing(future) => future,
+            State::Closed => return Poll::Ready(Err(get_closed_error())),
+        };
+
+        let result = match future.poll_unpin(cx) {
+            Poll::Ready(result) => result,
+            Poll::Pending => return Poll::Pending,
+        };
+
+        stream.state = State::Closed;
+        match result {
+            Ok(()) => Poll::Ready(Ok(())),
+            Err(error) => Poll::Ready(Err(error.into_futures_io_error())),
+        }
+    }
+}
+
+async fn write_bytes(
```

Comment:
I think it would be worth documenting here that the `u32` returned is number of chunks rather than number of bytes, which someone might assume without reading the implementation.